Move security guidance narratives to the security best practices page (#3099)
modelcontextprotocol/modelcontextprotocol: docs/docs/draft/tutorials/security/security_best_practices.mdx

properly issued _to the MCP server_ and passes them through to the
downstream API.

> This vulnerability has two critical dimensions:

Reviewer: On `main` this list lives in the [Access Token Privilege Restriction](https://modelcontextprotocol.io/specification/draft/basic/authorization/security-considerations#access-token-privilege-restriction) section, where "this vulnerability" is a server that "accepts tokens issued for other resources" and token passthrough is dimension two of it. Under the Token Passthrough heading here it reads as passthrough having a dimension called token passthrough. Carrying the antecedent over fixes it (and note `main` still has this text in the spec too, so it'll be in both places after the sync):

```suggestion
An attacker can gain unauthorized access or otherwise compromise an
MCP server if the server accepts tokens issued for other resources.
This vulnerability has two critical dimensions:
```

mitigates this by binding the response to the authorization server
the client recorded before redirecting, so the authorization code
cannot be redeemed at an unintended token endpoint. PKCE alone does
not prevent this attack because the client transmits the
`code_verifier` to the attacker's token endpoint. Resource indicators
do not help when the attacker's authorization server is intercepting
requests before they hit the honest authorization server. This
mitigation depends on honest authorization servers emitting `iss`; it
provides no protection against an honest server that does not.

### Localhost Redirect URI Impersonation

Native and locally-running MCP clients commonly use `localhost`
redirect URIs. When clients identify themselves with
[Client ID Metadata Documents](/specification/draft/basic/authorization/client-registration#client-id-metadata-documents),
the metadata document proves control of a domain, but it cannot prove
which local process is listening on a `localhost` redirect URI.

#### Attack Description

An attacker can claim to be any client by:

1. Providing the legitimate client's metadata URL as their `client_id`

> 2. Binding to the any `localhost` port, and providing that address as

Reviewer: Stray "the" carried over from the source text.

```suggestion
2. Binding to any `localhost` port, and providing that address as
```

   the redirect_uri
3. Receiving the authorization code via the redirect when the user
   approves

The server will see the legitimate client's metadata document and the
user will see the legitimate client's name, making attack detection
difficult.

#### Mitigation

See
[Localhost Redirect URI Risks](/specification/draft/basic/authorization/security-considerations#localhost-redirect-uri-risks)
in the authorization specification for the countermeasures expected
of authorization servers, including warning users when a `localhost`
redirect URI is in use and prominently displaying the hostname of

> non-localhost redirect URIs.

Reviewer: The [Localhost Redirect URI Risks](https://modelcontextprotocol.io/specification/draft/basic/authorization/security-considerations#localhost-redirect-uri-risks) list has an unconditional **MUST**, "clearly display the redirect URI hostname during authorization", plus a **SHOULD** for warnings on `localhost`-only redirect URIs. The "non-localhost" qualifier here narrows that MUST and exempts the exact case this section is about.

```suggestion
in the authorization specification for the countermeasures expected
of authorization servers, including displaying additional warnings for
`localhost`-only redirect URIs and clearly displaying the redirect URI
hostname during authorization.
```

  current operation along with related scopes that commonly work
  together, to reduce the number of step-up authorization rounds.
- **Extended approach**: Include the scopes required for the
  current operation, related scopes, and any other scopes the
  server anticipates the client may need in the near future.

The choice depends on the server's assessment of user experience impact and authorization friction.

Client guidance:

- Begin with only baseline scopes (or those specified by initial
  `WWW-Authenticate`)
- Cache recent failures to avoid repeated elevation loops for denied
  scopes

When the initial `WWW-Authenticate` challenge carries no `scope`
parameter, the
[Scope Selection Strategy](/specification/draft/basic/authorization#scope-selection-strategy)
directs clients to fall back to requesting all scopes listed in
`scopes_supported`. This approach accommodates the general-purpose
nature of MCP clients, which typically lack domain-specific knowledge
to make informed decisions about individual scope selection.
Requesting all available scopes allows the authorization server and
end-user to determine appropriate permissions during the consent
process.

> This approach minimizes user friction while following the principle of least privilege.

Reviewer: "This approach" shows up twice here and the second one is a stranded one-line paragraph left from the move. Folding it in:

```suggestion
Requesting all available scopes allows the authorization server and
end-user to determine appropriate permissions during the consent
process, minimizing user friction while following the principle of
least privilege.
```

<Note>
  Scope accumulation across operations is a client-side responsibility. Clients
  **SHOULD** compute the union of previously requested scopes and newly
  challenged scopes when initiating re-authorization, as described in [Step-Up
  Authorization
  Flow](/specification/draft/basic/authorization#step-up-authorization-flow).
  This allows servers to remain stateless with respect to client scope sets
  while ensuring clients do not lose previously granted permissions.
</Note>

<Note>
  **Hierarchical scopes**: Some authorization servers define scope hierarchies
  where a broader scope implies narrower ones (for example, an `admin` scope
  that subsumes `read`). When accumulating scopes, the client's union may

>   contain semantically redundant entries — for example, a token previously

Reviewer: Em dash here, splitting the sentence instead.

```suggestion
  contain semantically redundant entries. For example, a token previously
```
